cmd/deploy/input_test.go

```diff
@@ -2,9 +2,249 @@ package main
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": "localhost:5000/example:bar",
+			},
+			map[ociregistry.Digest]registry.Reference{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Tag:        "foo",
+					Digest:     "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				},
+				"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": {
+					Host:       "docker.io",
+					Repository: "foo/bar",
+					Tag:        "baz",
+					Digest:     "sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86",
+				},
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": {
+					Host:       "localhost:5000",
+					Repository: "example",
+					Tag:        "bar",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			"",
+			"",
+		},
+		{
+			"fallback",
+			map[string]string{
+				"": "example",
+			},
+			map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+				},
+			},
+			"",
+			"",
+		},
+		{
+			name: "bad digest",
+			rawLookups: map[string]string{
+				"invalid digest": "example:foo",
+			},
+			wantErr: `lookup key "invalid digest" invalid:`,
+		},
+		{
+			name: "bad reference",
+			rawLookups: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "://#",
+			},
+			wantErr: "failed to parse lookup ref",
+		},
+		{
+			name: "digest mistmatch",
+			rawLookups: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo@sha256:00f42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			},
+			wantErr: "digest on lookup ref should either be omitted or match key:",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			lookups, lookupDigest, err := normalizeInputLookup(x.rawLookups)
+
+			if x.wantErr != "" {
+				if err == nil {
+					t.Fatalf("Expected error not returned: %s", x.wantErr)
+				}
+
+				if !strings.Contains(err.Error(), x.wantErr) {
+					t.Fatalf("Expected error doesn't match. got:\n'%s',\n\nexpected to contain:\n'%s'", err.Error(), x.wantErr)
+				}
+			}
+
+			if x.wantDigest != "" {
+				if lookupDigest == nil {
+					t.Errorf("Expected last digest. got:\n%s\n\nexpected:\n%s\n", "<nil>", x.wantDigest)
+				} else if *lookupDigest != x.wantDigest {
+					t.Errorf("Expected last digest. got:\n%s\n\nexpected:\n%s\n", *lookupDigest, x.wantDigest)
+				}
+			}
+
+			for digest, ref := range lookups {
+				if _, ok := x.wantLookups[digest]; !ok {
+					t.Errorf("digest missing from lookups: %s", digest)
+				} else if ref != x.wantLookups[digest] {
+					t.Errorf("Reference doesn't match. got:\n%#v,\n\nexpected:\n%#v", ref, x.wantLookups[digest])
+				}
+			}
```

Comment:
Oh, this might also miss things (like returning too much, for example) -- I think you probably want to use https://pkg.go.dev/maps#Equal here :eyes: